src/nvim/mark_extended.h

```diff
@@ -0,0 +1,65 @@
+#ifndef NVIM_EXTMARK_H
+#define NVIM_EXTMARK_H
+
+#include "nvim/mark_extended_defs.h"
+#include "nvim/lib/kbtree.h"
+#include "nvim/lib/kvec.h"
+#include "nvim/map.h"
+
+typedef PMap(cstr_t) StringMap2; // TODO do these have to be seperate?
+
+typedef struct ExtendedMark ExtendedMark;
+struct ExtendedMark {
+  StringMap2 *names; // key = namespace, value = id
```

Comment:
Not sure, the common case I assume would be a position having just one or two marks, it seems wrong optimize specifically for the case of a position having plenty of marks. Also it seems semantically wrong: what if you have `"text[ns1:mark1]moretext[ns1:mark2]yetmore" `and delete `moretext`? now "ns1" would expect to find both mark1 and 2 at the same position, which this representation forbids.
For now, I would suggest just storing these as separate objects. If it _later_ turns out that multiple marks at the same position are common, we can try the solution: store the first (and maybe second) name directly in the same struct, allocate an array as soon as two(three) or more marks appears on the same position.